logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/ConversationRepository.kt

```diff
@@ -278,64 +280,89 @@ class ConversationDataSource(
-                    logMemberDetailsError(conversation, storageFailure)
-                    flowOf(Either.Left(storageFailure))
-                },
-                { otherUserId ->
-                    flowOf(otherUserId)
-                        .flatMapLatest { if (it != null) userRepository.getKnownUser(it) else flowOf(it) }
-                        .wrapStorageRequest()
-                        .map {
-                            it.map { otherUser ->
+        return observeConversationMembers(conversation.id)
+            .combine(userRepository.observeSelfUser()) { members, selfUser ->
+                selfUser to members.firstOrNull { item -> item.id != selfUser.id }
+            }
+            .flatMapLatest { (selfUser, otherMember) ->
+                val otherUserFlow = if (otherMember != null) userRepository.getKnownUser(otherMember.id) else flowOf(otherMember)
+                otherUserFlow
+                    .wrapStorageRequest()
+                    .mapRight { selfUser to it }
+            }
+            .flatMapLatest {
+                it.fold(
+                    { storageFailure ->
+                        logMemberDetailsError(conversation, storageFailure)
+                        flowOf(Either.Left(storageFailure))
+                    },
+                    { (selfUser, otherUser) ->
+                        combine(
+                            observeUnreadMessageCount(conversation),
+                            observeLastUnreadMessage(conversation),
+                            observeUnreadMentionsCount(conversation, selfUser.id)
+                        ) { unreadMessageCount: Long, lastUnreadMessage: Message?, unreadMentionsCount: Long ->
+                            Either.Right(
                                 conversationMapper.toConversationDetailsOneToOne(
                                     conversation = conversation,
                                     otherUser = otherUser,
                                     selfUser = selfUser,
-                                    unreadMessageCount = getUnreadMessageCount(conversation),
-                                    lastUnreadMessage = getLastUnreadMessage(conversation)
+                                    unreadMessageCount = unreadMessageCount,
+                                    unreadMentionsCount = unreadMentionsCount,
+                                    lastUnreadMessage = lastUnreadMessage
                                 )
-                            }
+                            )
                         }
-                }
-            )
+                    }
+                )
+            }
+    }
+
+    private suspend fun observeUnreadMessageCount(conversation: Conversation): Flow<Long> {
+        return if (conversation.supportsUnreadMessageCount && hasNewMessages(conversation)) {
+            messageDAO.observeUnreadMessageCount(idMapper.toDaoModel(conversation.id))
+        } else {
+            flowOf(0L)
```

Comment:
If a conversation as no new messages we return `flowOf(0L)`, what happens when a new message arrives in that conversation? Will we re-start the observation then or am I missing the context of where this is used?

Reply:
You're right, it won't work without a proper combination with `flatMap`, so I removed `hasNewMessages(conversation)`, this way it will always observe directly the DAO, the `if` with `conversation.supportsUnreadMessageCount` can stay because the type of the conversation can't be changed.